File: Athos/SeeDot/Codegen/EzPC.py
'''

Authors: Nishant Kumar.

Copyright:
Copyright (c) 2020 Microsoft Research
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

'''

import Util
import IR.IR as IR
import numpy as np
import Type as Type
import IR.IRUtil as IRUtil
from Codegen.CodegenBase import CodegenBase

class EzPC(CodegenBase):
	def __init__(self, writer, globalDecls, debugVar):
		self.out = writer
		self.globalDecls = globalDecls
		self.consSFUsed = Util.Config.consSF
		self.debugVar = debugVar

	def printAll(self, prog:IR.Prog, expr:IR.Expr):
		self._out_prefix()
		self.print(prog)
		self._out_suffix(expr)

	def _out_prefix(self):
		self.out.printf('\n\ndef void main(){\n')
		self.out.increaseIndent()
		self.printGlobalVarDecls()

	def printGlobalVarDecls(self):
		for decl in self.globalDecls:
			typ_str = IR.DataType.getIntStr()
			idf_str = decl
			declProp = self.globalDecls[decl]
			curType = declProp[0]
			if (len(declProp)>1):
				# If label specified in decl, then use that
				assert(len(declProp) >= 2 and len(declProp) <= 3) #For now only type, label and bitlen should be present
				variableLabel = ('pl' if (declProp[1] == 'public') else 'al')
				if (len(declProp) == 3):
					bitlen = declProp[2]
					typ_str = IR.DataType.getIntStrForBitlen(bitlen)
			else:
				# If variable unspecified, then default to secret
				variableLabel = 'al'
			if Type.isInt(curType): shape_str = ''
			elif Type.isTensor(curType): shape_str = ''.join(['[' + str(n) + ']' for n in curType.shape])
			self.out.printf('%s_%s%s %s;\n', typ_str, variableLabel, shape_str, idf_str, indent=True)
		self.out.printf('\n')

	def printFuncCall(self, ir:IR.FuncCall):
		self.out.printf("%s(" % ir.name, indent = True)
		keys = list(ir.argList)
		for i in range(len(keys)):
			arg = keys[i]
			self.print(arg)
			if i != len(keys) - 1:
				self.out.printf(", ")
		self.out.printf(");\n\n")

	def printForHeader(self, ir):
		assert(ir.endInt is not None and ir.endCond is None)
		self.out.printf('for ', indent=True)
		self.print(ir.var)
		self.out.printf(' = [%d: %d]{\n ', ir.st, ir.endInt)

	def printFor(self, ir):
		self.printForHeader(ir)
		self.out.increaseIndent()
		for cmd in ir.cmd_l:
			self.print(cmd)
		self.out.decreaseIndent()
		self.out.printf('};\n', indent=True)

	def printInt(self, ir:IR.Int):
		if (isinstance(ir.n, np.int32)):
			self.out.printf('%d', ir.n)
		elif (isinstance(ir.n, np.int64)):
			self.out.printf('%dL', ir.n)
		else:
			assert False

	def printInput(self, ir:IR.Input):
		if (ir.inputByParty==0):
			inputByPartyStr = "SERVER"
		elif (ir.inputByParty==1):
			inputByPartyStr = "CLIENT"
		else:
			assert(False) #For now the only supported values of party to input is 0 or 1
		self.out.printf('input({0}, {1}, '.format(inputByPartyStr, ir.expr.idf), indent=True)
		#assert(ir.dataType in ["DT_INT32"]) ####TODO: fix this
		if Util.Config.wordLength == 32:
			self.out.printf('int32_')
		elif Util.Config.wordLength == 64:
			self.out.printf('int64_')
		else:
			assert False
		if ir.isSecret:
			self.out.printf('al')
		else:
			self.out.printf('pl')
		for curDim in ir.shape:
			self.out.printf('[' + str(curDim) + ']')
		self.out.printf(');\n\n')

	def printComment(self, ir):
		self.out.printf('(* ' + ir.msg + ' *)\n', indent = True)

	def printDecl(self, ir):
		typ_str = IR.DataType.getIntStrForBitlen(ir.bitlen)
		variableLabel = 'pl' if not(ir.isSecret) else 'al'

		if Type.isInt(ir.typeExpr): shape_str = ''
		elif Type.isTensor(ir.typeExpr): shape_str = ''.join(['[' + str(n) + ']' for n in ir.typeExpr.shape])
		self.out.printf('%s_%s%s %s', typ_str, variableLabel, shape_str, ir.varIdf, indent=True)
		if (ir.value):
			assert(Type.isInt(ir.typeExpr)) #In EzPC ints can be declared and assigned in same line, not tensors
			self.out.printf(' = %s', str(ir.value[0]))
		self.out.printf(';\n\n')

	def _out_suffix(self, expr:IR.Expr):
		if self.debugVar is None:
			self.out.printf('output(CLIENT, ' + expr.idf + ');\n', indent=True)
		else:
			self.out.printf('output(CLIENT, ' + self.debugVar + ');\n', indent=True)
		self.out.decreaseIndent()
		self.out.printf('}\n', indent=True)
	

File: EzPC/EzPC/OBLIVC_example/ezpc.c
/*

Authors: Aseem Rastogi, Lohith Ravuru.

Copyright:
Copyright (c) 2020 Microsoft Research
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <obliv.h>
#include <obliv.oh>
#include "ezpc.h"
#include "../common/util.h"


int currentParty;

const char* mySide()
{
	if(currentParty==1) return "Generator";
	else return "Evaluator";
}

void flush_output_queue(protocolIO *io)
{   
	for (int i = 0; i < io->size ; ++i)
	{
		int role = io->outq[i].role;
		bool rolematch = false;
		rolematch = (role==0) ? true : ( (role == io->role) ? true : false ) ;
		if (rolematch)
		{
			uint32_t output = (io->outq[i]).output;
			fprintf(stderr,"%" PRIu32,output);
			fprintf(stderr,"\n");
		}
	}
	return;
}

uint32_t* add_to_output_queue(protocolIO *io, int role)
{
	(io->outq[io->size]).role = role;
	io->size = io->size + 1;
	return &((io->outq[(io->size) - 1]).output);
}

double lap;

int main(int argc, char *argv[])
{
	ProtocolDesc pd;
	int opt;
	opt=getopt(argc,argv,"r:");
	switch(opt){
	   case 'r':
		currentParty = atoi(optarg);
		currentParty += 1; 
		break;
	   default:
		printf("Usage: ./a.out -r <0|1> \n");
		exit(0);
	}
	if(currentParty!=1 && currentParty!=2)
	{	
		printf("Usage: ./a.out -r <0|1> \n");
		exit(0);
	}
	const char* remote_host = currentParty==1? "localhost" : NULL;
	ocTestUtilTcpOrDie(&pd,remote_host,"1234");
	// scanf("%d",&(io.myinput));

	setCurrentParty(&pd,currentParty);
	lap=wallClock();
	protocolIO io;
	io.role = currentParty;
	io.size = 0;
	execYaoProtocol(&pd,ezpc_main,&io);
	flush_output_queue(&io);
	fprintf(stderr,"%s total time: %lf s\n",mySide(),wallClock()-lap);
	fprintf(stderr,"Yao gate count: %u\n",io.gatecount);
	cleanupProtocol(&pd);
	return 0;
}


obliv uint32_t uarshift(obliv uint32_t a, uint32_t b) obliv {
	obliv uint32_t result;
	obliv if(a > (0 - a))
	{
		result = (0 - ((0 - a) >> b));
	}
	else
	{
		result = (a >> b);
	}
	return result;
}
